usl_models/usl_models/flood_ml/prediction_dataset.py

```diff
@@ -0,0 +1,195 @@
+
+    This dataset produces the input for `model.call_n`.
+    For training with teacher-forcing, `load_dataset_windowed` should be used instead.
+    The examples are generated from multiple simulations.
+    The dataset iteratively yields examples read from Google Cloud Storage to avoid
+    pulling all examples into memory at once.
+
+    Args:
+      
+      batch_size: Size of batches yielded by the dataset. Approximate memory
+                  usage is 10GB * batch_size during training.
+      n_flood_maps: The number of flood maps in each example.
+      m_rainfall: The width of the temporal rainfall tensor.
+      max_chunks: The maximum number of examples to yield from the dataset.
+                  If `None` (default) yields all examples from the simulations.
+      firestore_client: The client to use when interacting with Firestore.
+      storage_client: The client to use when interacting with Cloud Storage.
+    """
+    firestore_client = firestore_client or firestore.Client()
+    storage_client = storage_client or storage.Client()
+
+    def generator():
+        """Generator for producing full inputs from study area"""
+        for model_input, metadata in _iter_model_inputs_for_prediction(
+        firestore_client,
+        storage_client,
+        city_cat_config,
+        study_area,
+        n_flood_maps,
+        m_rainfall,
+        max_chunks
+        ):
+            yield model_input, metadata
+
+    # Create the dataset for this simulation
+    dataset = tf.data.Dataset.from_generator(
+        generator=generator,
+        output_signature=(
+            dict(
+                geospatial=tf.TensorSpec(
+                    shape=(
+                        constants.MAP_HEIGHT,
+                        constants.MAP_WIDTH,
+                        constants.GEO_FEATURES,
+                    ),
+                    dtype=tf.float32,
+                ),
+                temporal=tf.TensorSpec(
+                    shape=(constants.MAX_RAINFALL_DURATION, m_rainfall),
+                    dtype=tf.float32,
+                ),
+                spatiotemporal=tf.TensorSpec(
+                    shape=(
+                        n_flood_maps,
+                        constants.MAP_HEIGHT,
+                        constants.MAP_WIDTH,
+                        1,
+                    ),
+                    dtype=tf.float32,
+                ),
+            ),
+            dict(
+                feature_chunk = tf.TensorSpec(shape=(), dtype=tf.string),
+                rainfall= tf.TensorSpec(shape=(), dtype=tf.int32),
+            ),
+            ),
+          )
+    # If no batch specified, do not batch the dataset, which is required
+    # for generating data for batch prediction in VertexAI.
+    if batch_size:
+        print("batch: ", batch_size)
+        dataset = dataset.batch(batch_size)
+    return dataset
+
+def _generate_temporal_tensor(
```

Comment:
A lot of this code looks copy-pasted from dataset.py
We should refactor the common parts of the code used for building prediction & training datasets. It looks like the primary difference is whether we're calling

get_spatial_feature_chunk_metadata_for_prediction
vs
get_spatial_feature_chunk_metadata

or
 
get_temporal_feature_metadata_for_prediction
vs
get_temporal_feature_metadata

and the code should be otherwise nearly identical.

Reply:
Perhaps a better API is to pass a bool arg `training=False` for prediction case, which is pretty common in Keras libs.

Reply:
I will look into refactoring this.

Reply:
@waltaskew how would you like to me refactor this? The prediction functions takes `city_cat_config` name for retrieving temporal tensors and the `study_area` for feature tensors.

Reply:
It looks like there is a fair amount of copy-pasted code between the two. A small example -- both have copies of `_download_as_array` and `_download_as_tensor`. The two files should be importing and calling the same function rather than having duplicate copies of the function.
Generally, the more duplicated logic between the two files can be factored into shared functions used by both, the better.

Reply:
Ah, yes, i can change that for sure - my bad on that one. Is there any suggestion on the main functions?